docs/api/remix.md

```diff
@@ -1295,6 +1295,211 @@ return new Response("", {
 });
 ```
 
+## `parseMultipartFormData` (node)
+
+Allows you to handle multipart forms (file uploads) for your app.
+
+Would be useful to understand [the Browser File API](https://developer.mozilla.org/en-US/docs/Web/API/File) to know how to use this API.
+
+It's to be used in place of `request.formData()`.
+
+```diff
+- let formData = await request.formData();
++ let formData = await parseMultipartFormData({ request, uploadHandler });
+```
+
+For example:
+
+```tsx [2-5,7,23]
+export let action: ActionFunction = async ({ request }) => {
+  let formData = await parseMultipartFormData({
+    request,
+    uploadHandler // <-- we'll look at this deeper next
+  });
+
+  // the returned value for the file field is whatever our uploadHandler returns.
+  // Let's imagine we're uploading the avatar to s3,
+  // so our uploadHandler returns the URL.
+  let avatarUrl = formData.get("avatar");
+
+  // update the currently logged in user's avatar in our database
+  await updateUserAvatar(request, avatarUrl);
+
+  // success! Redirect to account page
+  return redirect("/account");
+};
+
+export default function AvatarUploadRoute() {
+  return (
+    <Form method="post">
+      <label htmlFor="avatar-input">Avatar</label>
+      <input id="avatar-input" type="file" name="avatar" />
+      <button>Upload</button>
+    </Form>
+  );
+}
+```
+
+### `uploadHandler`
+
+The `uploadHandler` is the key to whole thing. It's responsible for what happens to the file as it's being streamed from the client. You can save it disk, store it in memory, or act as a proxy to send it somewhere else (like a file storage provider).
+
+Remix has two utilities to create `uploadHandler`s for you:
+
+- `createFileUploadHandler`
+- `createMemoryUploadHandler`
+
+These are fully featured utilities for handling fairly simple use cases. It's not recommended to load anything but quite small files into memory. Saving files to disk is a reasonable solution for many use cases. But if you want to upload the file to a file hosting provider, then you'll need to write your own.
+
+#### `createFileUploadHandler`
+
+**Example:**
+
+```tsx
+let uploadHandler = createFileUploadHandler({
+  directory: "public/uploads",
+  maxFileSize: 5_000_000,
+  file: ({ filename }) => filename
+});
+
+export let action: ActionFunction = async ({ request }) => {
+  let formData = await parseMultipartFormData({
+    request,
+    uploadHandler
+  });
+
+  let file = formData.get("avatar");
+
+  // file is a "NodeFile" which has a similar API to "File"
```

Comment:
Why is `file` of type `NodeFile` here if we're saying it's of type `File` with a memoryHandler and it should be `File`, `string` or `undefined` when creating a custom handler?

> Your job is to do whatever you need with the `stream` and return a value that's a valid [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) value: [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File), `string`, or `undefined`.

Reply:
NodeFile is an implementation of the File interface that keeps things out of memory and on disk until you need it. It is infact a valid FormData value :)